shared/utility/math/filter/ParticleFilter.hpp

```diff
@@ -38,242 +72,396 @@ namespace utility::math::filter {
+                            return stats::resample::residual(model.n_particles,
+                                                             weights.begin(),
+                                                             weights.end(),
+                                                             stats::resample::systematic<decltype(weights.begin())>);
+                        default: throw std::runtime_error("Invalid setting for residual method."); break;
+                    }
+                    break;
+                case ResampleMethod::STRATIFIED:
+                    return stats::resample::stratified(model.n_particles, weights.begin(), weights.end());
+                case ResampleMethod::SYSTEMATIC:
+                    return stats::resample::systematic(model.n_particles, weights.begin(), weights.end());
+                default: throw std::runtime_error("Invalid setting for resample method."); break;
             }
         }
 
-        void set_state(const std::vector<StateVec>& initial_mean, const std::vector<StateMat>& initialCovariance) {
-            particles =
-                sample_particles(initial_mean, initialCovariance, model.getParticleCount() + model.getRogueCount());
-
-            // Limit the state of each particle to ensure they are still valid
-            for (unsigned int i = 0; i < particles.cols(); i++) {
-                particles.col(i) = model.limit(particles.col(i));
+        /**
+         * @brief Do a weighted resampling of all of the particles in the filter. Resampled particles are then perturbed
+         * (using model::noise) and the model is asked to provide new rogue particles (if the model is using
+         * them). Resampled particles are then limited (using model::limit) to ensure validity and all particle weights
+         * are reset to 1.
+         *
+         * @param dt The amount of time that has elapsed since the previous time update step.
+         */
+        void resample(const Scalar& dt) {
+
+            // Make a multivariate normal distribution with zero mean and the model's process noise as the covariance
+            // This will be used to perturb the resampled particles
+            MultivariateNormal<Scalar, Model::size> multivariate(model.noise(dt));
+
+            // Get indexes to the particles that are being resampled
+            // Some particles may be resampled multiple times
+            std::vector<int> idx = resample_particles();
+
+            // Resample the particles
+            // For each index get the corresponding particle and perturb it with the model's process noise
+            ParticleList resampled_particles(Model::size, particles.cols());
+            for (int i = 0; i < int(idx.size()); ++i) {
+                // Create a new particle by using the selected particle as the mean + value from process noise
+                resampled_particles.col(i) = particles.col(idx[i]) + multivariate.sample();
```

Comment:
Might be a good idea to do this perturbing of the state by the multivariate and the limiting in the time update rather than here. It's not really part of resampling and it will make it clearer each step that we are doing